cpp/src/sampling/detail/sampling_utils_impl.cuh

```diff
@@ -448,5 +631,320 @@ sample_edges(
+                                                       minors.begin(),
+                                                       edge_ids->begin(),
+                                                       hops->begin(),
+                                                       labels->begin()),
+                             h_tx_value_counts,
+                             handle.get_stream());
+          } else {
+            std::forward_as_tuple(std::tie(majors, minors, edge_ids, labels), std::ignore) =
+              shuffle_values(comm,
+                             thrust::make_zip_iterator(
+                               majors.begin(), minors.begin(), edge_ids->begin(), labels->begin()),
+                             h_tx_value_counts,
+                             handle.get_stream());
+          }
+        }
+      } else {
+        if (edge_types) {
+          if (hops) {
+            std::forward_as_tuple(std::tie(majors, minors, edge_types, hops, labels), std::ignore) =
+              shuffle_values(comm,
+                             thrust::make_zip_iterator(majors.begin(),
+                                                       minors.begin(),
+                                                       edge_types->begin(),
+                                                       hops->begin(),
+                                                       labels->begin()),
+                             h_tx_value_counts,
+                             handle.get_stream());
+          } else {
+            std::forward_as_tuple(std::tie(majors, minors, edge_types, labels), std::ignore) =
+              shuffle_values(
+                comm,
+                thrust::make_zip_iterator(
+                  majors.begin(), minors.begin(), edge_types->begin(), labels->begin()),
+                h_tx_value_counts,
+                handle.get_stream());
+          }
+        } else {
+          if (hops) {
+            std::forward_as_tuple(std::tie(majors, minors, hops, labels), std::ignore) =
+              shuffle_values(comm,
+                             thrust::make_zip_iterator(
+                               majors.begin(), minors.begin(), hops->begin(), labels->begin()),
+                             h_tx_value_counts,
+                             handle.get_stream());
+          } else {
+            std::forward_as_tuple(std::tie(majors, minors, labels), std::ignore) = shuffle_values(
+              comm,
+              thrust::make_zip_iterator(majors.begin(), minors.begin(), labels->begin()),
+              h_tx_value_counts,
+              handle.get_stream());
+          }
+        }
+      }
+    }
+  }
+
+  return std::make_tuple(std::move(majors),
+                         std::move(minors),
+                         std::move(weights),
+                         std::move(edge_ids),
+                         std::move(edge_types),
+                         std::move(hops),
+                         std::move(labels));
```

Comment:
And this function does not sort the output based on labels (and I guess this is the right design), so we need to sort the output based on labels before constructing an offset array for the labels.